### PR TITLE
Crime rate

### PR DESCRIPTION
**Pull request by jonjunduan**
Add this crime rate dataset. 
1. get data from StatsCan
2. calculate the weighted DA level crime rate using the Econ team's lookup table which links RESP (police region) to DA. Population is used as a weight when different RESPs cover the same DAs.

### CODE REVIEW

File: src/03_output_crime_rate.R [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+# Use Case
+# get_cansim_sqlite is particularly beneficial when working with large tables or when you need to perform complex filtering operations before bringing data into memory. For example, table 35-10-0184-01 is several gigabytes in size, making it impractical to load entirely into memory.
+# However, for smaller tables that can be easily processed in memory, get_cansim may still be a simpler and more straightforward option. The choice between the two functions depends on the specific requirements of your data analysis task and the size of the dataset you're working with.
+# it was supposed to save sqlite to a cache folder. 
+cansim_id <- "35-10-0184-01"
+# it is too slow to index in sqlite on a network drive, so switch to a local folder, and create a copy from LAN use_network_path("data/cansim_cache") to C drive repo_folder/data. 
+options(cansim.cache_path = "./data")
+# options(cansim.cache_path = use_network_path("data/cansim_cache"))
+getOption("cansim.cache_path")
+connection <- cansim::get_cansim_sqlite(cansim_id,
```
Comment (thister) [inline src/03_output_crime_rate.R:68]: I can't proceed!

```
Error in `.data$cansimTableNumber`:
! Column `cansimTableNumber` not found in `.data`.
Run `rlang::last_trace()` to see where the error occurred.
Warning messages:
1: In dir.create(cache_path) :
  cannot create dir '.\data\cansim_35100184_eng', reason 'No such file or directory'
2: In dir.create(base_dir) :
  cannot create dir '.\data\cansim_35100184_eng', reason 'No such file or directory'
```

Comment (jonjunduan) [reply src/03_output_crime_rate.R:68]: We need to copy the cached file from the LAN to the c: drive local repo folder. It is a huge file so it is impossible to work on it when it stays in the LAN directly.

Comment (thister) [reply src/03_output_crime_rate.R:68]: Ahhh...so that's a step I just have to do myself beforehand? Interesting.

Comment (thister) [reply src/03_output_crime_rate.R:68]: Would it make sense to add this line then?

```
fs::file_copy("G:/Operations/Data Science and Analytics/2024 SES Index/data/cansim_cache/35100184-eng.sqlite", "./data/")
```

Comment (thister) [reply src/03_output_crime_rate.R:68]: (But use the `safepaths` way)

Comment (jonjunduan) [reply src/03_output_crime_rate.R:68]: > Would it make sense to add this line then?
> 
> ```
> fs::file_copy("G:/Operations/Data Science and Analytics/2024 SES Index/data/cansim_cache/35100184-eng.sqlite", "./data/")
> ```

Wow. This is great idea with `safepaths` way.

File: src/03_output_crime_rate.R [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+
+##########################################################################
+# For CSD or DAs, many CSDs or DAs have to share RESP together, so it is better to calculate the ratios within each RESP and CSDs or DAs could share the ratios. 
+# If one CD or CSD has two or more RESPs, we could average them weighting by the number of the postal code regions within the RESPs. 
+# Econ team provides a lookup table for us to link DA to RESP weighted by postal code or population
+###########################################################################
+
+
+DA_RESP_lookup = read_excel(path = use_network_path("data/crime_rate/Pop by DA and RESP.xlsx"),
```
Comment (thister) [inline src/03_output_crime_rate.R:164]: Similar problem as before. This works:

```DA_RESP_lookup = readxl::read_excel(path = "G:/Operations/Data Science and Analytics/2024 SES Index/data/crime_rate/Pop by DA and RESP.xlsx", sheet = "DA RESP")
```

I'm still not sure what the best way to do this is - .Renviron file?

File: src/03_output_crime_rate.R [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+                             sheet = "DA RESP")
+
+DA_RESP_lookup %>%
+  count( DA_2021 )
+# 3,712
+# some DAs cover multiple RESP
+DA_RESP_lookup %>%
+  count(RESP,)
+# # 195
+
```
Comment (thister) [inline src/03_output_crime_rate.R:174]: The first row has two "NULL"s -- filter them out?

**Review by thister — COMMENTED** [review]
I wasn't 100% sure what you are doing at the bottom with all those joins but it seems to work

File: src/03_output_crime_rate.R [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+  group_by(ref_date,violations,statistics,DA_2021) %>% 
+  summarise(value = weighted.mean(value, w = POP_CNT))
+
+bc_da_crime_stats_year_weighted_by_pop %>% write.csv2( here::here("out/BC_DA_Crime_Rate_DIP.csv"))
+# bc_da_crime_stats_year_weighted_by_pc = bc_da_crime_stats_year %>% 
+#   group_by(ref_date,violations,statistics,DA_2021) %>% 
+#   summarise(value = weighted.mean(value, w = PC_CNT))
+
+# later, we may calculate the moving average of the crime rate instead of using the observed rate. 
+
```
[the remark this answers is not in the document]
Comment (jonjunduan) [reply src/03_output_crime_rate.R:203]: two levels: 1. Rate per 100,000 population; 2.Percentage change in rate

File: src/03_output_crime_rate.R (hunk 2 above)
Comment (jonjunduan) [reply src/03_output_crime_rate.R:164]: I see. We need to set_network_path first. `config.yml` or .Renviron may work better.

File: src/03_output_crime_rate.R [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+
+##############################################################
+# Data Dictionary
+#############################################################
+
+crime_rate_dict_labels = c(
+  "ref_date" = "The year of the observation (in '%Y' format)" ,
+  "violations"  = "Violation type and classification, such as ",
```
[the remark this answers is not in the document]
Comment (jonjunduan) [reply src/03_output_crime_rate.R:212]: Oh. I forgot to fill in those placeholders.

File: src/03_output_crime_rate.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonjunduan) [reply src/03_output_crime_rate.R:203]: I create a table with all possible years, violation types, and statistics type combinations using across_join. There may be a better way to do it, like a grid function. Then this table is used to join the weighted crime rate table calculated using StatsCan's table.

[the remark this answers is not in the document]
Comment (thister) [reply src/03_output_crime_rate.R:203]: oh haha you're right - my mistake.

File: src/03_output_crime_rate.R (hunk 1 above)
Comment (jonjunduan) [reply src/03_output_crime_rate.R:68]: I have added it.

File: src/03_output_crime_rate.R (hunk 2 above)
Comment (jonjunduan) [reply src/03_output_crime_rate.R:164]: Use the safepaths::set_network_path function to set up the network path first, and it should work.

File: src/03_output_crime_rate.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonjunduan) [reply src/03_output_crime_rate.R:212]: Done. Add examples of the violations.

File: src/03_output_crime_rate.R (hunk 3 above)
Comment (jonjunduan) [reply src/03_output_crime_rate.R:174]: Right. That is for the BC totaL. I filtered them out.

**Pull request merged**